src/tests/ftest/util/AgentUtils.py

```diff
@@ -127,31 +127,32 @@ def run_agent(basepath, server_list, client_list=None):
         sessions[client] = p
 
     # double check agent launched successfully
-    file_desc = sessions[client].stderr.fileno()
-    flags = fcntl.fcntl(file_desc, fcntl.F_GETFL)
-    fcntl.fcntl(file_desc, fcntl.F_SETFL, flags | os.O_NONBLOCK)
-    timeout = 5
-    start_time = time.time()
-    result = 0
-    pattern = "Starting daos_agent"
-    expected_data = ""
-    while True:
-        output = ""
-        try:
-            output = sessions[client].stderr.read()
-        except IOError as excpn:
-            if excpn.errno != errno.EAGAIN:
-                raise excpn
-            continue
-        match = re.findall(pattern, output)
-        expected_data += output
-        result += len(match)
-        if not output or time.time() - start_time > timeout:
-            print("<AGENT>: {}".format(expected_data))
-            raise AgentFailed("DAOS Agent didn't start!")
-        break
-    print("<AGENT> agent started and took %s seconds to start" % \
-            (time.time() - start_time))
+    for client in client_list:
+        file_desc = sessions[client].stderr.fileno()
+        flags = fcntl.fcntl(file_desc, fcntl.F_GETFL)
+        fcntl.fcntl(file_desc, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+        timeout = 5
+        start_time = time.time()
+        result = 0
+        pattern = "Starting daos_agent"
+        expected_data = ""
+        while True:
+            output = ""
+            try:
+                output = sessions[client].stderr.read()
+            except IOError as excpn:
+                if excpn.errno != errno.EAGAIN:
+                    raise excpn
+                continue
+            match = re.findall(pattern, output)
```

Comment:
The code doesn't appear to be confirming that the "Starting daos_agent" string appears in the output.  Could there also be a premature exit of the loop if the daos_agent does not print anything into stderr between loops?  Might I suggest replacing lines 147-155 with the following:

            expected_data += output
            if time.time() - start_time > timeout:
                print("<AGENT>: {}".format(expected_data))
                raise AgentFailed("DAOS Agent didn't start!")
            else:
                match = re.findall(pattern, output)
                if len(match) > 0:
                    print("<AGENT> agent started on node {} in {} seconds".format(client, time.time() - start_time))
                    break

Reply:
agreed, except for the timeout part, that needs to be checked above the continue statement, most easily in the while condition.

Reply:
Switched original snippet for Phil's suggestion, with timeout moved as Scott suggested. It was still left as a discrete check, see other comment.